File: packages/platform-mobile/android/app/src/main/java/app/bramble/mobile/Totp.kt
package app.bramble.mobile

import android.net.Uri
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

// Computes a live TOTP code from a stored key, for one-time-code autofill (parity with the
// iOS ProvidesOneTimeCodes path). Mirrors util/totp.ts: accepts an `otpauth://totp/...`
// URI or a bare base32 secret; RFC 6238 defaults (SHA1, 6 digits, 30s). Only the digits
// ever leave here, never the seed. Returns null for HOTP / garbage so callers can skip it.
object Totp {

    fun generate(key: String?, nowMillis: Long): String? {
        val raw = key?.trim()
        if (raw.isNullOrEmpty()) return null
        return try {
            var secretB32 = raw
            var digits = 6
            var period = 30L
            var algorithm = "SHA1"
            if (raw.startsWith("otpauth://", ignoreCase = true)) {
                if (!raw.startsWith("otpauth://totp", ignoreCase = true)) return null // HOTP unsupported
                val uri = Uri.parse(raw)
                secretB32 = uri.getQueryParameter("secret") ?: return null
                uri.getQueryParameter("digits")?.toIntOrNull()?.let { digits = it }
                uri.getQueryParameter("period")?.toLongOrNull()?.let { period = it }
                uri.getQueryParameter("algorithm")?.let { algorithm = it.uppercase() }
            }
            val secret = base32Decode(secretB32) ?: return null
            val counter = (nowMillis / 1000L) / period
            hotp(secret, counter, digits, macAlgorithm(algorithm))
        } catch (e: Exception) {
            null
        }
    }

    private fun macAlgorithm(algorithm: String): String = when (algorithm) {
        "SHA256" -> "HmacSHA256"
        "SHA512" -> "HmacSHA512"
        else -> "HmacSHA1"
    }

    private fun hotp(secret: ByteArray, counter: Long, digits: Int, mac: String): String {
        val msg = ByteArray(8)
        var c = counter
        for (i in 7 downTo 0) {
            msg[i] = (c and 0xff).toByte()
            c = c shr 8
        }
        val hmac = Mac.getInstance(mac).apply { init(SecretKeySpec(secret, mac)) }.doFinal(msg)
        val offset = hmac[hmac.size - 1].toInt() and 0x0f
        val binary = ((hmac[offset].toInt() and 0x7f) shl 24) or
            ((hmac[offset + 1].toInt() and 0xff) shl 16) or
            ((hmac[offset + 2].toInt() and 0xff) shl 8) or
            (hmac[offset + 3].toInt() and 0xff)
        val mod = Math.pow(10.0, digits.toDouble()).toInt()
        return (binary % mod).toString().padStart(digits, '0')
    }

    private fun base32Decode(input: String): ByteArray? {
        val clean = input.replace("[\\s-]".toRegex(), "").trimEnd('=').uppercase()
        if (clean.isEmpty()) return null
        val alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
        var buffer = 0
        var bitsLeft = 0
        val out = ArrayList<Byte>(clean.length * 5 / 8)
        for (ch in clean) {
            val v = alphabet.indexOf(ch)
            if (v < 0) return null
            buffer = (buffer shl 5) or v
            bitsLeft += 5
            if (bitsLeft >= 8) {
                bitsLeft -= 8
                out.add(((buffer shr bitsLeft) and 0xff).toByte())
            }
        }
        return out.toByteArray()
    }
}

File: packages/platform-mobile/ios/App/AutoFillProbe/Totp.swift
import CryptoKit
import Foundation

// Computes a live TOTP code from a stored key, for one-time-code AutoFill (ProvidesOneTimeCodes).
// Mirrors android/.../Totp.kt and util/totp.ts: accepts an `otpauth://totp/...` URI or a bare
// base32 secret; RFC 6238 defaults (SHA1, 6 digits, 30s). Only the digits ever leave here, never
// the seed. Returns nil for HOTP / garbage so callers can skip the entry.
enum Totp {

	static func generate(_ key: String?, now: Date = Date()) -> String? {
		guard let raw = key?.trimmingCharacters(in: .whitespacesAndNewlines), !raw.isEmpty else {
			return nil
		}
		var secretB32 = raw
		var digits = 6
		var period = 30.0
		var algorithm = "SHA1"
		if raw.lowercased().hasPrefix("otpauth://") {
			guard raw.lowercased().hasPrefix("otpauth://totp"),  // HOTP unsupported
				let items = URLComponents(string: raw)?.queryItems,
				let secret = items.first(where: { $0.name == "secret" })?.value
			else { return nil }
			secretB32 = secret
			if let d = items.first(where: { $0.name == "digits" })?.value, let v = Int(d) { digits = v }
			if let p = items.first(where: { $0.name == "period" })?.value, let v = Double(p) { period = v }
			if let a = items.first(where: { $0.name == "algorithm" })?.value { algorithm = a.uppercased() }
		}
		// The dynamic truncation yields 31 bits, so more than 10 digits is meaningless.
		guard period > 0, (1...10).contains(digits), let secret = base32Decode(secretB32) else {
			return nil
		}
		let counter = UInt64(max(0, (now.timeIntervalSince1970 / period).rounded(.down)))
		return hotp(secret: secret, counter: counter, digits: digits, algorithm: algorithm)
	}

	private static func hotp(secret: Data, counter: UInt64, digits: Int, algorithm: String) -> String?
	{
		var be = counter.bigEndian
		let msg = withUnsafeBytes(of: &be) { Data($0) }
		let key = SymmetricKey(data: secret)
		let mac: Data
		switch algorithm {
		case "SHA256": mac = Data(HMAC<SHA256>.authenticationCode(for: msg, using: key))
		case "SHA512": mac = Data(HMAC<SHA512>.authenticationCode(for: msg, using: key))
		default: mac = Data(HMAC<Insecure.SHA1>.authenticationCode(for: msg, using: key))
		}
		guard let last = mac.last else { return nil }
		let offset = Int(last & 0x0f)
		guard mac.count >= offset + 4 else { return nil }
		let binary =
			(UInt32(mac[offset] & 0x7f) << 24) | (UInt32(mac[offset + 1]) << 16)
			| (UInt32(mac[offset + 2]) << 8) | UInt32(mac[offset + 3])
		let mod = UInt32(pow(10.0, Double(digits)))
		return String(format: "%0\(digits)u", binary % mod)
	}

	private static func base32Decode(_ input: String) -> Data? {
		var clean = input
			.replacingOccurrences(of: "[\\s-]", with: "", options: .regularExpression)
			.uppercased()
		while clean.hasSuffix("=") { clean.removeLast() }
		if clean.isEmpty { return nil }
		let alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567")
		var buffer = 0
		var bitsLeft = 0
		var out = Data()
		for ch in clean {
			guard let v = alphabet.firstIndex(of: ch) else { return nil }
			buffer = (buffer << 5) | v
			bitsLeft += 5
			if bitsLeft >= 8 {
				bitsLeft -= 8
				out.append(UInt8((buffer >> bitsLeft) & 0xff))
			}
		}
		return out.isEmpty ? nil : out
	}
}
